README.md
# Cloudflare Workers example using Durable Objects and Split SDK

Based on [Durable Objects TypeScript Rollup ES Modules template](https://github.com/cloudflare/durable-objects-typescript-rollup-esm), this is a NPM package template for kick starting a [Cloudflare Worker project](https://developers.cloudflare.com/workers/) using [Split SDK](https://help.split.io/hc/en-us/articles/360058730852-Browser-SDK) with [Durable Objects](https://developers.cloudflare.com/workers/learning/using-durable-objects).

## Overview

In Cloudflare Workers, an Split SDK factory instance must be created on each incoming HTTP request where we want to evaluate features. However, by default, each instance fetches the rollout plan (i.e., feature flag and segment definitions) it needs to compute treatments from Split backend, impacting the response latency.

To reduce the latency, this project stores the rollout plan in a [Durable Object](https://developers.cloudflare.com/workers/learning/using-durable-objects), a low-latency and consistent storage hosted on the same Cloudflare Workers infrastructure, and instructs the SDKs to "consume" data from that storage instead of fetching it from Split backend.

The project overall architecture is ilustrated in the following diagram:

<p align="center">
  <img alt="Overview" src="./diagram.png" width="720">
</p>

It has the following modules:

- [`src/SplitStorage.ts`](./src/SplitStorage.ts): durable object class that will be used as a low-latency storage for your rollout plans.
- [`src/index.ts`](./src/index.ts): entrypoint of your Cloudflare Worker script, where incoming HTTP requests and scheduled requests are handled.
  - Incoming HTTP requests (see [Fetch Event](https://developers.cloudflare.com/workers/runtime-apis/fetch-event)) are external requests for which we provide some logic. Here the Split SDK is instantiated in "partial consumer" mode (see [Sharing state with a pluggable storage](https://help.split.io/hc/en-us/articles/360058730852-Browser-SDK#sharing-state-with-a-pluggable-storage)). In this mode, features are evaluated by reading the rollout plan from the `SplitStorage` durable object instead of fetching it from Split backend. But unlike "consumer" mode, the SDK posts impressions and events directly to Split backend instead of storing them.
  - Scheduled requests (see [Scheduled Event](https://developers.cloudflare.com/workers/runtime-apis/scheduled-event)) are requests scheduled by a CRON trigger to periodically execute the [Split Javascript Synchronizer](https://help.split.io/hc/en-us/articles/4421513571469-Split-JavaScript-synchronizer-tools), the module in charge of storing and updating the rollout plan in a `SplitStorage` instance.
- [`src/SplitStorageWrapper.ts`](./src/SplitStorageWrapper.ts): adapter used by both the Split SDK and Synchronizer which offers all the methods that the Split packages will need to read and write data from a given storage implementation, in this case the `SplitStorage` durable object instance.

## Setup

1. Install [Wrangler CLI](https://developers.cloudflare.com/workers/cli-wrangler):

```bash
npm install -g @cloudflare/wrangler
```

2. Scaffold a Cloudflare Workers project from this GitHub repository:

```bash
wrangler generate [your-project-name] https://github.com/splitio/cloudflare-workers-template
```

3. Run `npm install` to install dependencies.

4. Update your `account_id` in the `wrangler.toml` file. You can get it from the Cloudlfare dashboard or running `wrangler dev`.

5. Update your `sdkKey` in `src/index.ts` with your Split server-side SDK key. You can get it from the [Split user interface](https://help.split.io/hc/en-us/articles/360019916211-API-keys).

6. Deploy your worker, locally with `wrangler dev` or in Cloudflare with `wrangler publish`.

### Testing your deployment

To test your deployment, first synchronize the Split Storage with the data of your rollout plan (feature flag and segment definitions). For that, either wait the CRON trigger to execute, or manually perform a requests to `/sync` to instruct the worker to synchronize data.

```bash
curl 'https://<YOUR-WORKER-DOMAIN>/sync'
// Synchronization success
```

Then you can send a request to `/get-treatment`, which simply evaluates a given feature flag with a given user key, provided as query params, and returns a treatment.

```bash
curl 'https://<YOUR-WORKER-DOMAIN>/get-treatment?key=<SOME-USER-ID>&feature-flag=<SOME-FEATURE-FLAG-NAME>'
// Treatment: on
```

**Troubleshooting**: if you get a 500 Internal Server Error from your worker, check that `SplitStorage` durable object binding is properly configured in your wrangler.toml file. Take into account that Durable Objects are only available with a Workers paid subscription.


CHANGES.txt
1.2.0 (December 12, 2023)
 - Updated Javascript Sync Tools to version 0.5.0 and Browser SDK to version 0.13.0, which include support for flag sets in consumer modes.

1.1.2 (June 30, 2023)
 - Updated Javascript Browser SDK dependency.

1.1.1 (May 15, 2023)
 - Updated terminology on the SDKs codebase to be more aligned with current standard without causing a breaking change. The core change is the term split for feature flag on things like logs and IntelliSense comments.
 - Updated Javascript Sync Tools and Browser SDK dependencies.

1.1.0 (October 5, 2022)
 - Updated Javascript Sync Tools and Browser SDK dependencies, which include support for OPTIMIZED and NONE impression modes.
 - Updated dependencies for vulnerability fixes.

1.0.1 (July 12, 2022)
 - Updated dependencies for vulnerability fixes.

1.0.0 (January 12, 2022)
 - Initial public release.


src/SplitStorage.ts
/**
 * Durable object to use as Split Storage
 */
export class SplitStorage {
  // https://developers.cloudflare.com/workers/runtime-apis/durable-objects#transactional-storage-api
  storage: DurableObjectStorage;

  constructor(state: DurableObjectState) {
    this.storage = state.storage!;
  }

  // Handle HTTP requests from clients.
  async fetch(request: Request) {
    let url = new URL(request.url);
    let reqBody, result;
    let param = url.searchParams.get("param");

    switch (url.pathname) {
      /** Key-Value operations */

      case "/get":
        // Setting allowConcurrency to improve performance. By default, the system will prevent concurrent events from executing while awaiting a read operation
        result =
          (await this.storage.get<string>(param!, {
            allowConcurrency: true
          })) || null;
        break;

      case "/set":
        reqBody = await request.json<string>();
        // There is no need to await write operations. Any series of write operations with no intervening await will automatically be submitted atomically
        this.storage.put(param!, reqBody);
        break;

      case "/getAndSet":
        // Using default allowConcurrency (false), to guarantee atomic operation.
        // A series of reads followed by a series of writes are automatically atomic and behave like a transaction
        const getResult = this.storage.get<string>(param!);
        reqBody = await request.json<string>();
        this.storage.put(param!, reqBody);
        result = (await getResult) || null;
        break;

      case "/del":
        this.storage.delete(param!);
        break;

      case "/getKeysByPrefix":
        result = await this.storage.list<string>({
          prefix: param!,
          allowConcurrency: true
        });
        result = Array.from(result.keys());
        break;

      case "/getMany":
        const keys = param!.split(",");
        const map = await this.storage.get<string>(keys, {
          allowConcurrency: true
        });
        result = keys.map(key => map.get(key) || null);
        break;

      case "/incr":
        reqBody = await request.json<number>();
        result = await this.storage.get<number>(param!); // allowConcurrency false
        result = result ? result + reqBody : reqBody;
        this.storage.put(param!, result);
        break;

      case "/decr":
        reqBody = await request.json<number>();
        result = await this.storage.get<number>(param!); // allowConcurrency false
        result = result ? result - reqBody : -reqBody;
        this.storage.put(param!, result);
        break;

      /** Set operations */

      case "/itemContains":
        reqBody = await request.json<string>();
        result = await this.storage.get<Set<string>>(param!, {
          allowConcurrency: true
        });
        result = result && result.has(reqBody) ? true : false;
        break;

      case "/addItems":
        reqBody = await request.json<string[]>();
        result =
          (await this.storage.get<Set<string>>(param!)) || new Set<string>(); // allowConcurrency false
        for (let i = 0; i < reqBody.length; i++) result.add(reqBody[i]);
        this.storage.put(param!, result);
        break;

      case "/removeItems":
        reqBody = await request.json<string[]>();
        result = await this.storage.get<Set<string>>(param!); // allowConcurrency false
        if (result) {
          for (let i = 0; i < reqBody.length; i++) result.delete(reqBody[i]);
          this.storage.put(param!, result);
        }
        break;

      case "/getItems":
        result = await this.storage.get<Set<string>>(param!, {
          allowConcurrency: true
        });
        result = Array.from(result || new Set<string>());
        break;

      /** Queue operations */
      // Since SDK must run in partial consumer mode, events and impressions are
      // not tracked and so there is no need to implement Queue operations

      case "/pushItems":
        break;

      case "/popItems":
        result = new Array<string>();
        break;

      case "/getItemsCount":
        result = 0;
        break;

      // Extra endpoint to clear the storage, for testing or fixing issues when migrating the durable object instance
      case "/deleteAll":
        await this.storage.deleteAll();
        break;

      default:
        return new Response("Not found", { status: 404 });
    }

    return new Response(
      result !== undefined ? JSON.stringify(result) : undefined
    );
  }
}


src/SplitStorageWrapper.ts
/**
 * Creates a storage wrapper for a given Split Storage Durable Object
 * https://developers.cloudflare.com/workers/runtime-apis/durable-objects#object-stubs
 *
 * @param {DurableObjectStub} durableObject durable object stub
 * @returns {import("@splitsoftware/splitio-commons/src/storages/types").IPluggableStorageWrapper} storage wrapper
 */
export function SplitStorageWrapper(durableObject: DurableObjectStub) {
  function doFetch(path: string, param: string, param2?: any) {
    // For now, fetch requires a valid URL. So we have to provide a dummy URL that will be ignored at the other end
    // See https://github.com/cloudflare/workers-chat-demo/blob/master/src/chat.mjs#L518
    return durableObject.fetch(
      `https://dummy-url/${path}?param=${param}`,
      param2 && {
        method: "POST",
        body: JSON.stringify(param2)
      }
    );
  }

  return {
    /** Key-Value operations */

    /**
     * Get the value of given `key`.
     *
     * @function get
     * @param {string} key Item to retrieve
     * @returns {Promise<string | null>} A promise that resolves with the element value associated with the specified `key`,
     * or null if the key does not exist.
     */
    async get(key: string) {
      return (await doFetch("get", key)).json<string>();
    },

    /**
     * Add or update an item with a specified `key` and `value`.
     *
     * @function set
     * @param {string} key Item to update
     * @param {string} value Value to set
     * @returns {Promise<boolean>} A promise that resolves if the operation success, whether the key was added or updated.
     */
    async set(key: string, value: string) {
      return (await doFetch("set", key, value)).ok;
    },

    /**
     * Add or update an item with a specified `key` and `value`.
     *
     * @function getAndSet
     * @param {string} key Item to update
     * @param {string} value Value to set
     * @returns {Promise<string | null>} A promise that resolves with the previous value associated to the given `key`, or null if not set.
     */
    async getAndSet(key: string, value: string) {
      return (await doFetch("getAndSet", key, value)).json<string>();
    },

    /**
     * Removes the specified item by `key`.
     *
     * @function del
     * @param {string} key Item to delete
     * @returns {Promise<boolean>} A promise that resolves if the operation success, whether the key existed and was removed or it didn't exist.
     */
    async del(key: string) {
      return (await doFetch("del", key)).ok;
    },

    /**
     * Returns all keys matching the given prefix.
     *
     * @function getKeysByPrefix
     * @param {string} prefix String prefix to match
     * @returns {Promise<string[]>} A promise that resolves with the list of keys that match the given `prefix`.
     */
    async getKeysByPrefix(prefix: string) {
      return (await doFetch("getKeysByPrefix", prefix)).json<string[]>();
    },

    /**
     * Returns the values of all given `keys`.
     *
     * @function getMany
     * @param {string[]} keys List of keys to retrieve
     * @returns {Promise<(string | null)[]>} A promise that resolves with the list of items associated with the specified list of `keys`.
     * For every key that does not hold a string value or does not exist, null is returned.
     */
    async getMany(keys: string[]) {
      return (await doFetch("getMany", keys.join(","))).json<string[]>();
    },

    /** Integer operations */

    /**
     * Increments the number stored at `key` by `increment`, or set it to `increment` if the value doesn't exist.
     *
     * @function incr
     * @param {string} key Key to increment
     * @param {number} increment Value to increment by. Defaults to 1.
     * @returns {Promise<number>} A promise that resolves with the value of key after the increment.
     */
    async incr(key: string, increment = 1) {
      return (await doFetch("incr", key, increment)).json<number>();
    },

    /**
     * Decrements the number stored at `key` by `decrement`, or set it to minus `decrement` if the value doesn't exist.
     *
     * @function decr
     * @param {string} key Key to decrement
     * @param {number} decrement Value to decrement by. Defaults to 1.
     * @returns {Promise<number>} A promise that resolves with the value of key after the decrement.
     */
    async decr(key: string, decrement = 1) {
      return (await doFetch("decr", key, decrement)).json<number>();
    },

    /** Set operations */

    /**
     * Returns if item is a member of a set.
     *
     * @function itemContains
     * @param {string} key Set key
     * @param {string} item Item value
     * @returns {Promise<boolean>} A promise that resolves with true boolean value if `item` is a member of the set stored at `key`,
     * or false if it is not a member or `key` set does not exist.
     */
    async itemContains(key: string, item: string) {
      return (await doFetch("itemContains", key, item)).json<boolean>();
    },

    /**
     * Add the specified `items` to the set stored at `key`. Those items that are already part of the set are ignored.
     * If key does not exist, an empty set is created before adding the items.
     *
     * @function addItems
     * @param {string} key Set key
     * @param {string} items Items to add
     * @returns {Promise<boolean | void>} A promise that resolves if the operation success.
     */
    async addItems(key: string, items: string[]) {
      await doFetch("addItems", key, items);
    },

    /**
     * Remove the specified `items` from the set stored at `key`. Those items that are not part of the set are ignored.
     *
     * @function removeItems
     * @param {string} key Set key
     * @param {string} items Items to remove
     * @returns {Promise<boolean | void>} A promise that resolves if the operation success. If key does not exist, the promise also resolves.
     */
    async removeItems(key: string, items: string[]) {
      await doFetch("removeItems", key, items);
    },

    /**
     * Returns all the items of the `key` set.
     *
     * @function getItems
     * @param {string} key Set key
     * @returns {Promise<string[]>} A promise that resolves with the list of items. If key does not exist, the result is an empty list.
     */
    async getItems(key: string) {
      return (await doFetch("getItems", key)).json<string[]>();
    },

    // No-op. No need to connect to DurableObject stub
    async connect() {
      // Never throws if Durable Object namespace binding is properly configured
      if (!durableObject) throw new Error("Durable Object stub not provided");
    },

    // No-op. No need to disconnect from DurableObject stub
    async disconnect() {},

    /** Queue operations */
    // Since Split SDK must run in partial consumer mode, events and impressions are
    // not tracked and so there is no need to implement Queue operations

    async pushItems(key: string, items: string[]) {},

    async popItems(key: string, count: number) {
      return [];
    },

    async getItemsCount(key: string) {
      return 0;
    }
  };
}


src/index.ts
import { Synchronizer } from "@splitsoftware/splitio-sync-tools";
import {
  SplitFactory,
  PluggableStorage,
  ErrorLogger
} from "@splitsoftware/splitio-browserjs";
import { SplitStorageWrapper } from "./SplitStorageWrapper";
// In order for the workers runtime to find the class that implements
// our Durable Object namespace, we must export it from the root module.
export { SplitStorage } from "./SplitStorage";

/**
 * Worker script.
 *
 * Implemented using the new "Module Worker" syntax (See https://developers.cloudflare.com/workers/learning/migrating-to-module-workers).
 * This new syntax is required when using Durable Objects, because they are implemented by classes, and those classes need to be exported.
 *
 * To be deployed, this worker must be configured with a Durable Object namespace bindings call 'SplitStorage', which is already configured
 * in the wrangler.toml file. Take into account that Durable Objects are only available with a Workers paid subscription.
 */
export default {
  // Handle HTTP incoming requests
  async fetch(request: Request, env: Env) {
    const url = new URL(request.url);

    switch (url.pathname) {
      // Use Split SDK to evaluate a feature flag. Request example `/get-treatment?key=some_key&feature-flag=some_feature_flag_name`
      case "/get-treatment":
        return handleGetTreatmentRequest(url, env);

      // Synchronize your Split Storage, same as scheduled requests
      case "/sync":
        return handleSynchronization(env);

      // Clean up all data from your Split Storage. SDK instances will evaluate to 'control'
      case "/delete-all":
        const durableObjectStub = getSplitStorage(env);
        return durableObjectStub.fetch("https://dummy-url/deleteAll");

      default:
        return new Response("Not found", { status: 404 });
    }
  },

  // Handle scheduled requests by a cron trigger, use to synchronize your Split Storage periodically
  async scheduled(
    controller: ScheduledController,
    env: Env,
    ctx: ExecutionContext
  ) {
    return handleSynchronization(env);
  }
};

interface Env {
  SplitStorage: DurableObjectNamespace;
}

// Server-side SDK key is required for Synchronizer and sufficient for JS Browser SDK
const sdkKey = "<YOUR-SERVER-SIDE-SDK-KEY>";

// Get reference to Split Storage durable object
function getSplitStorage(env: Env) {
  // Here we use Split SDK key as durable object name, but any name can be used.
  // Actually, multiple SDKs with different SDK keys could access the same durable object,
  // as long as they set different storage prefixes to avoid data collisions.
  const id = env.SplitStorage.idFromName(sdkKey);
  return env.SplitStorage.get(id);
}

// Use Split SDK to evaluate a feature flag
async function handleGetTreatmentRequest(url: URL, env: Env) {
  const key = url.searchParams.get("key");
  if (!key) return new Response("No key provided", { status: 400 });

  const featureFlagName = url.searchParams.get("feature-flag");
  if (!featureFlagName)
    return new Response("No feature flag name provided", { status: 400 });

  // SDK instances are created in 'consumer_partial' mode, which access
  // the Split Storage to get the rollout plan data for evaluations
  const factory = SplitFactory({
    core: {
      authorizationKey: sdkKey,
      key
    },
    mode: "consumer_partial",
    storage: PluggableStorage({
      wrapper: SplitStorageWrapper(getSplitStorage(env))
    }),
    // Disable or keep only ERROR log level in production, to minimize performance impact
    debug: ErrorLogger()
  });
  const client = factory.client();

  // Await until the SDK is ready or timed out, for which treatment evaluations will be 'control'.
  // Timed out should never happen if SplitStorage durable object binding is properly configured.
  await new Promise(res => {
    client.on(client.Event.SDK_READY, res);
    client.on(client.Event.SDK_READY_TIMED_OUT, res);
  });

  // Async evaluation, because it access the rollout plan from the Split Storage
  const treatment = await client.getTreatment(featureFlagName);

  // Flush data to Split backend. But not await, in order to reduce response latency
  client.destroy();

  // Do something with the treatment
  return new Response(`Treatment: ${treatment}`);
}

// Synchronize rollout plan data into Split Storage durable object
async function handleSynchronization(env: Env) {
  const synchronizer = new Synchronizer({
    core: {
      authorizationKey: sdkKey
    },
    storage: {
      type: "PLUGGABLE",
      wrapper: SplitStorageWrapper(getSplitStorage(env))
    },
    // Disable or keep only ERROR log level in production, to minimize performance impact
    debug: "ERROR"
  });

  try {
    await new Promise<void>((res, rej) => {
      synchronizer.execute(error => (error ? rej(error) : res()));
    });
    return new Response("Synchronization finished");
  } catch (error) {
    return new Response(`Synchronization failed with error: ${error}`, {
      status: 500
    });
  }
}
